Add FCIOPutState

ssailer opened the pull request:
Refactor individual put logic into static inline functions so they can be reused in a new `FCIOPutState`.

Reserve three new tags:
```
FCIOEventHeader = 7
FCIOFSPConfig = 8
FCIOFSPEvent = 9
FCIOFSPStatus = 10
```

These are not know by fcio, but used externally, however to avoid conflicts their tags are reserved here.
This leaves the option to add them later if needed.

Functions which write/read `FCIOEventHeader` have been added.
These should write the `type,pulser,timestamp,timeoffset,deadregion,trace_list` fields and their corresponding sizes (if applicable) and the `theader` data.

fwerner commented on src/fcio.h:
```diff
@@ -226,7 +226,11 @@ typedef enum {
-  FCIOSparseEvent = 6
+  FCIOSparseEvent = 6,
+  FCIOEventHeader = 7,
+  FCIOFSPConfig = 8,
```
Perhaps add some comment that these three tags are reserved for the Fast Stream Processor.

ssailer replied:
Addressed

fwerner commented on tests/fcio_test_unknown_tags.c:
```diff
@@ -14,6 +14,8 @@
 
 int main(int argc, char* argv[])
 {
+  assert(argc >= 2);
```
Haven't checked if other tests do the same - but we should aim to avoid assert for functional checks.

fwerner commented on src/fcio.c:
```diff
@@ -334,7 +334,11 @@ typedef enum {
-  FCIOSparseEvent = 6
+  FCIOSparseEvent = 6,
+  FCIOEventHeader = 7,
+  FCIOFSPConfig = 8,
```
Same here - add some comment.

fwerner commented on src/fcio.c:
```diff
@@ -449,6 +453,25 @@ Returns the internal FCIOStream object on success or NULL on error.
 }
 
+
+static inline int fcio_put_config(FCIOStream output, fcio_config* config)
```
The fcio_put functions do not perform any input checks - we can either accept that and be very careful in their use (because they are static functions) or we add just two cheap checks `if (!output || !config) return` to be defensive.

ssailer replied:
Added

fwerner commented on src/fcio.c:
```diff
@@ -449,6 +453,25 @@ Returns the internal FCIOStream object on success or NULL on error.
+
 /*=== Function ===================================================*/
 
 int FCIOPutConfig(FCIOStream output, FCIOData *input)
```
As mentioned above this (and all functions below) needs an additional null-pointer check on `input`.

ssailer replied:
Added

fwerner commented on src/fcio.c:
```diff
@@ -657,19 +740,22 @@ Returns the return value of the individual FICOPut functions or 0 on unknown tag
 {
   switch (tag) {
```
Same here - need to check `output` and `input` before calling `fcio_put_*`.

fwerner commented on src/fcio.c:
```diff
@@ -763,23 +849,55 @@ static inline void fcio_get_sparseevent(FCIOStream stream, fcio_event *event, in
+
+  const int length = config->eventsamples + 2;
+  unsigned short read_buffer[FCIOMaxChannels * 2];
+  FCIOReadUShorts(stream, event->num_traces * 2, read_buffer);
```
Out-of-bounds if `num_traces` (which is not checked above) is > `FCIOMaxChannels * 2`.

fwerner commented on src/fcio.c:
```diff
@@ -763,23 +849,55 @@ static inline void fcio_get_sparseevent(FCIOStream stream, fcio_event *event, in
+  event->timestamp_size = FCIOReadInts(stream,10,event->timestamp)/sizeof(int);
+  event->deadregion_size = FCIOReadInts(stream,10,event->deadregion)/sizeof(int);
+  FCIOReadInts(stream,1,&event->num_traces);
+  FCIOReadUShorts(stream,FCIOMaxChannels,event->trace_list);
```
Need to check the return value here to safely iterate over the trace list below.

fwerner commented on src/fcio.c:
```diff
@@ -1600,6 +1743,37 @@ FCIOState *FCIOGetNextState(FCIOStateReader *reader, int *timedout)
+/*--- Description ------------------------------------------------//
+
+//----------------------------------------------------------------*/
+{
```
Need to check output and state before dereference and calling `fcio_put_*`.

fwerner reviewed (commented):
I left some in-line comments, mostly regarding safe/defensive programming.

ssailer commented:
Tried to address these comments in for new commits:

- Added description for FCIOTag enum, marks FSP tags as reserved
- Add new test to check record write/read consistency, needs more work
- Add requrest input guards. static inline functions received some as well, should be thrown out for the `PutRecord` style functions by the compiler, but kept for the `PutState` function. Introduced return values for `get` functions, of which at least errors are passed on.
- Introduced additional checks for records during reading - would enjoy comments. Might have inadvertently put one in the previous commit.

fwerner reviewed (commented):
Wow - what a set of changes! Looks much better with the homogeneous return values. Just a couple of inline comments left to address. 👍

ssailer commented on src/fcio.c:
```diff
@@ -494,27 +536,32 @@ A record consist of the message tag and all data members of the struct.
+
+static inline int fcio_put_event(FCIOStream output, fcio_config* config, fcio_event* event)
+{
+  if (!output || !config || !event) return 0;
```
Yes.

ssailer commented on src/fcio.c:
```diff
@@ -530,22 +577,38 @@ to remote depends on their corresponding *_size items.
+
+static inline int fcio_put_sparseevent(FCIOStream output, fcio_config* config, fcio_event* event)
+{
+  if (!output || !config || !event) return 0;
```
yes.

ssailer commented on src/fcio.c:
```diff
@@ -565,33 +628,85 @@ to remote depends on their corresponding *_size items.
 
 //----------------------------------------------------------------*/
 {
   if (!output){
```
Fixed.

ssailer commented on src/fcio.c:
```diff
@@ -651,140 +754,248 @@ valid record tags are described above
+
   FCIOReadInt(stream,config->adcs);
   FCIOReadInt(stream,config->triggers);
+  const int num_configured_channels = config->adcs + config->triggers;
```
In particular case the `num_configured_channels` is only checked against the `FCIOReadInts` later in `fcio_get_config` to emit a warning. Not all versions of f250b write the `tracemap` field afaik.

Except for `fcio_get_sparsevent` all calls to array reading `FCIORead` wrappers use the maximum allocated array length to read data and emit warnings if they don't match expectations. At least that was the intention. :)

ssailer commented on src/fcio.c:
```diff
@@ -651,140 +754,248 @@ valid record tags are described above
+       fprintf(stderr,"FCIO/fcio_get_config/DEBUG: trace %d mapped to 0x%x\n",i,config->tracemap[i]);
+  }
+
+  // if tracemap size does not match, there is probable
```
yes.

fwerner commented:
LGTM! 👍
